blobby/src/decode.rs

```diff
@@ -0,0 +1,271 @@
+use super::{Error, NEXT_MASK, VAL_MASK};
+
+pub(crate) const fn read_vlq(data: &mut &[u8]) -> Result<usize, Error> {
+    let b = match data.split_first() {
+        Some((&b, rest)) => {
+            *data = rest;
+            b
+        }
+        None => return Err(Error::UnexpectedEnd),
+    };
+    let mut next = b & NEXT_MASK;
+    let mut val = (b & VAL_MASK) as usize;
+
+    macro_rules! step {
+        () => {
+            if next == 0 {
+                return Ok(val);
+            }
+            let b = match data.split_first() {
+                Some((&b, rest)) => {
+                    *data = rest;
+                    b
+                }
+                None => return Err(Error::UnexpectedEnd),
+            };
+
+            next = b & NEXT_MASK;
+            let t = (b & VAL_MASK) as usize;
+            val = ((val + 1) << 7) + t;
+        };
+    }
+
+    step!();
+    step!();
+    step!();
+
+    if next != 0 {
+        return Err(Error::InvalidVlq);
+    }
+
+    Ok(val)
+}
+
+macro_rules! try_read_vlq {
+    ($data:expr) => {
+        match read_vlq(&mut $data) {
+            Ok(v) => v,
+            Err(err) => return Err(err),
+        }
+    };
+}
+
+pub const fn parse_dedup_len(mut data: &[u8]) -> Result<usize, Error> {
+    read_vlq(&mut data)
+}
+
+pub const fn parse_items_len(mut data: &[u8]) -> Result<usize, Error> {
```

Comment:
Unfortunately, to get number of items stored in a blb file we have to fully parse it. In other words, `parse_into_slice` effectively parses the file twice, which may be noticeably slow on large enough files. We could resolve this by adding number of stored items to the start of the file, but it would mean that we have to re-generate our existing blb files.